File: 152-maximum-product-subarray/maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int pre = 1;
        int suf = 1;
        int n = nums.length;
        int ans = Integer.MIN_VALUE;
        for(int i = 0; i<n; i++){
            if(suf == 0) suf=1;
            if(pre == 0) pre=1;
            pre *= nums[i];
            suf *= nums[n-i-1];
            ans = Math.max(ans, Math.max(pre, suf));
        }
        return ans;
    }
}    

File: 152-maximum-product-subarray/maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        maxi = nums[0]
        mini = nums[0]
        result = nums[0]
        for i in range(1, len(nums)):
            temp = maxi
            maxi = max(nums[i], max(nums[i]*maxi, nums[i]*mini))
            mini = min(nums[i], min(nums[i]*temp, nums[i]*mini))
            result = max(result, maxi)
        return result
        